### DISCUSSION TITLE
Typed property hydration doesn't support `null` values

### DISCUSSION THREAD

**Discussion by jameslkingsley** [livewire/livewire#4537] [7. Bugs (archived)]
There exists a lot of support for various property types such as `Wireable` and `Carbon`, however these only work if the initial value is an instance of one. In a real world scenario, this is unlikely to be the case.

Imagine properties for start and end dates. Initially these would be `null`, and only become date instances when the user picks them. Right now there is no way to support this natively without having to hack it together with update hooks.

In the `HydratePublicProperties` middleware where it determines the property type, instead of comparing the value type, it should be comparing the property type as well using Reflection. I believe this would then properly support typed properties with a `null` initial value.

Maybe #4154 already takes care of this to some degree? If not then maybe it's worth a PR?

**Comment by joshhanley** [comment]
@jameslkingsley so I'm pretty sure Caleb doesn't like to use type hints, and they aren't enforced in php anyway, so it needs to be written in a way that supports both typed properties and non-typed properties.

Yeah the goal of #4154 is that you can create your own handlers, so you can then decide what happens in those instances when the property is initially null and then an update happens.

Hope this helps!

**Reply by jameslkingsley** [reply]
@joshhanley Yeah I was thinking that it would first check the value type (as that's the cheapest), and if that's not a match, then check if the property has a declared type (if not then do whatever it does now). If the property does have a declared type — then resolve its hydration handler to hydrate/dehyrate it.

Hopefully #4154 adds the tools to make this work, and then the default type hydrations in Livewire could be refactored to support a lot more things.

Tbh I'm not sure how far you can go without type-hints in terms of support in Livewire — without them you're always comparing the incoming value type with no discernible logic for how that type should be handled. For example a simple `public $date` could receive a string of `"2022-01-18"` but you may want a `Carbon` instance instead. Which I guess makes sense — no type hints = get whatever JavaScript gives you — type hints = let Livewire cast to your chosen type.

**Reply by joshhanley** [reply]
@jameslkingsley yeah exactly, that's the hope!

Yeah when there's no typehints that's where the hooks have to come into play so you explicitly convert that string to a carbon instance.
